[iOS] AWSAPIPlugin GraphQL Advanced workflow docs with custom GraphQL requests (#2141)
aws-amplify/docs: docs/lib/graphqlapi/fragments/ios/advanced-workflows.md

This section describes different use cases for constructing your own custom GraphQL requests and how to approach it. You may want to construct your own GraphQL request if you want to
- retrieve only a subset of the data to reduce data transfer
- retrieve nested objects at a depth that you choose
- combine multiple operations into a single request

> - decode the GraphQL response to a format of your choice

Reviewer: Since Amplify is an opinionated framework, we should avoid giving customers a "choice" of response format.    I do get that you need to explain what `responseType` and `decodePath` should be, but I would just remove this bullet point.

Author: removing

A GraphQL request is automatically generated for you when using AWSAPIPlugin with the existing workflow. For example, if you have a Todo model, a mutation request to save the Todo will look like this:
```swift
let todo = Todo(name: "my first todo", description: "todo description")
Amplify.API.mutate(request: .create(todo))
```
Underneath the covers, a request is generated with a GraphQL document and variables and sent to the AppSync service. 

```json

> // GraphQL document

Author: sure, updating

## Subset of data

The selection set of the document specifies which fields are returned in the response. For example, If you are displaying a view of the Todo without the description, you can construct the document to omit the field. You can learn more about the selection sets [here](https://spec.graphql.org/draft/#sec-Selection-Sets).

```
query getTodo($id: ID!) {
  getTodo(id: $id) {
    id
    name
  }
}
```
The response data will look like this
```json
{
  "data": {
    "getTodo": {
      "id": "111",
      "name": "my first todo"
    }
  }
}
```
First, create your own `GraphQLRequest`

```swift
extension GraphQLRequest {
    static func getWithoutDescription(byId id: String) -> GraphQLRequest<Todo> {
        let operationName = "getTodo"
        let document = """
        query getTodo($id: ID!) {
          \(operationName)(id: $id) {
            id
            name
          }
        }
        """
        return GraphQLRequest<Todo>(document: document,
                                    variables: ["id": id],
                                    responseType: Todo.self,
                                    decodePath: operationName)
    }
}
```

> The decode path specifies which part of the response to deserialize to the `responseType`. You'll need to specify the operation name to deserialize the object at "data.getTodo" successfully into a Todo model.

Author: thanks for the insight, just got something working on iOS for a PaginatedList: https://github.com/aws-amplify/amplify-ios/pull/715

> The decode path specifies which part of the response to deserialize to the `responseType`. You'll need to specify the operation name to deserialize the object at "data.getTodo" successfully into a Todo model.

Author: In terms of the decodePath (https://github.com/aws-amplify/amplify-ios/pull/714) we do want the change to have a default behavior of assuming data in an object with single child, however this does break the API behavior for existing customers (see 3rd point in customer impact in PR), so we decided to hold off on releasing this in 1.x.x versions. What i've done in this docs PR is to consolidate the two sections, and remove decodePath from the Android sections `20_custom.md` in iOS contains decodePath and not in Android

```graphql
enum PostStatus {
  ACTIVE
  INACTIVE
}
type Post @model {
  id: ID!
  title: String!
  rating: Int!
  status: PostStatus!
  comments: [Comment] @connection(keyName: "byPost", fields: ["id"])
}
type Comment @model
  @key(name: "byPost", fields: ["postID", "content"]) {
  id: ID!
  postID: ID!
  post: Post! @connection(fields: ["postID"])
  content: String!
}
```
The `GraphQLRequest` will look like this

```swift
extension GraphQLRequest {
    static func getPostWithComments(byId id: String) -> GraphQLRequest<JSONValue> {
        let document = """
        query getPost($id: ID!) {
          getPost(id: $id) {
            id
            title
            rating
            status
            comments {
              items {
                id
                postID
                content
              }
            }
          }
        }
        """
        return GraphQLRequest<JSONValue>(document: document,
                                         variables: ["id": id],
                                         responseType: JSONValue.self)
    }
}

```

> Query with `Amplify.API.query(request: .getCommentWithPost(byId: "[COMMENT_ID]"))`. Notice here that `JSONValue` is used as the `responseType`. `JSONValue` is utility type that can be used to represent an arbitrary JSON response.

Reviewer: Is there not a way for the customer to get a strongly typed response (i.e. a Post object and a list of Comment objects)?  For Android, the deserializer supports this, with an unlimited depth.

Author: The issue is first, the Model-to-GraphQLRequest builder does not generate the selection set of the comments so data is not returned. Even if we default to a depth of 2 and do add the selection set for comments, then the second issue is when deserializing it doesn't match the response data structure. The response data contains the post fields, followed by the "items" key, and inside "items" is the array of Comments. For a successfull deserialization, the type should be 
````
Post {
  var items: [Comment]
}
``` 
instead of 
Post {
  var comments: [Comment]
}
```
Does Android check for "items" and do some custom deserialization into the Post's list of comments model? Even if it does do a default selection set of depth 2, the list of comments should be paginated, how does the developer get the next x number of comments for that post?

Reviewer: > The issue is first, the Model-to-GraphQLRequest builder does not generate the selection set of the comments so data is not returned. 

On Android, this is not an issue since we default to a depth of 2, for the API category.    To change the depth, all we need to do is expose a way to provide that constant.

> 
> Does Android check for "items" and do some custom deserialization into the Post's list of comments model?

Yes.  The expected type is `List<Comment>`, which is processed by the [`IterableDeserializer`](  https://github.com/aws-amplify/amplify-android/blob/27503d7a64c014e39eea4062d1962fb5b5128a27/aws-api/src/main/java/com/amplifyframework/api/aws/IterableDeserializer.java).  If the `JsonElement` is a `JsonObject`, rather than a `JsonArray`, (it is, and it contains `items` and `nextToken`), then the deserializer parses the value of `items` as a `List<Comment>`.

> Even if it does do a default selection set of depth 2, the list of comments should be paginated, how does the developer get the next x number of comments for that post?

Great question - Android doesn't support pagination on nested models, since we discard the pagination details during deserialization.   It is a limitation, but I don't think it's a huge deal because customers could still obtain a page of Comments by making a new request with Comments as the top level item, which does support pagination.

Author: So it is clear that Android and IOS contains different walk-depths by default, and it will take some major verions of the API to change this behavior. What i've done here is for the "nested data" section, instead of showcasing a walk-depth of two (this would be odd for Android since the default is already 2), it's showcasing the scenario where there's two nested objects and developer only wants to retrieve one of them. Post contains list of comments and and list of notes and only wants to get post with comments, omitting the notes

## Combining Multiple Operations

When you want to perform more than one operation in a single request, you can place them within the same document. For example, to retrieve a Post and a Todo
```swift
extension GraphQLRequest {
    static func get(byPostId postId: String, todoId: String) -> GraphQLRequest<JSONValue> {
        let document = """
        query get($postId: ID!, $todoId: ID!) {
          getPost(id: $postId) {
            id
            title
            rating
          }
          getTodo(id: $todoId) {
            id
            name
          }
        }
        """
        return GraphQLRequest<JSONValue>(document: document,
                                         variables: ["postId": postId,
                                                     "todoId": todoId],
                                         responseType: JSONValue.self)
    }
}
```

> Once you have the response data in a `JSONValue`, you can access each object in the JSON structure by encoding it back to Data and decoding it to the expected Model.

Reviewer: For Android, there is not a way to do this currently.   The deserializer will throw an exception if there is more than one element in the `data` object.    To add support for this, we could either:
1. If `data` has one element, deserialize into a `responseType` object (like we already do), but if more than one, deserialize into a `Map<{operationName}, {responseType}>`.
2. Add support for `decodePath` similar to iOS.

I'm inclined towards #1 because I'd prefer to avoid requiring customers to specify a `decodePath`.  I'm open to thoughts from you, because I think we should have platform parity.

Author: For 1.  I agree that when data only has one element, then the developer should be able to just specify the response type without having to set the decodePath to the operationName. the `decodePath` parameter isn't very useful when the data contains arrays, then you won't be able to go further than the objects before that, for example,
```
data {
  getTodo {
    id: 111
    items [
       {  id: 222 }, { id: 333 }
    ]
  }
  getPost {
    id: 444
  }
}
```
decodePath of "getTodo.items" is the furthest you can go since the value at "items" is an array of objects

when there is more than one, can we take no opinion on deserialization and just deserialize to whatever the response is at "data"? this will be an object like
```
ResponseType {
  operationName1: ResponseType1
  operationName2: ResponseType2
}
```
and would map to the same response structure that the developer can see if they use the console or tried deserialization to String and printed it out

Reviewer: Ahh, that's right - #1 isn't actually possible since we would need the customer to provide the expected `responseType` for each operation, and currently, the GraphQLRequest constructor only asks for one `responseType`.

So, there's no way to support multiple operations with Android, currently. 

I know there has been at least one customer request for it, so I will keep thinking about the best way to support this, but for now, I think we should just say it is not supported in the docs.

Author: on iOS, we can first get parity in behavior: "if there is only one element at data, deserialize to it" as API behavior change, this will allow the documentation to be updated to omit the decodePath. 

for elements which is under "data.items", we'll need custom deserialization logic in the plugin code to support this like Android has. 

until both of these changes happen, APIPlugin on iOS requires the decodePath to be set for both scenarios
1. getting element under "data"
2. getting list of elements under "data.items" for list queries

Author: This PR on iOS should address the decodePath parity and then I can update the docs here to remove the `decodePath` examples https://github.com/aws-amplify/amplify-ios/pull/714

Author: getting list of elements under "data.items" has been merged in, however decodePath will go in later

```swift
Amplify.API.query(request: .get(byPostId: "[POST_ID]", todoId: "[TODO_ID]")) { result in
    switch result {
    case .success(let response):
        switch response {
        case .success(let data):
            if let todoJSON = data.value(at: "getTodo"),
                let todoData = try? JSONEncoder().encode(todoJSON),
                let todo = try? JSONDecoder().decode(Todo.self, from: todoData) {
                print(todo)
            }
            if let postJSON = data.value(at: "getPost"),
                let postData = try? JSONEncoder().encode(postJSON),
                let post = try? JSONDecoder().decode(Post.self, from: postData) {
                print(post)
            }
        case .failure(let errorResponse):
            print("Response contained errors: \(errorResponse)")
        }
    case .failure(let apiError):
        print("Failed with error: \(apiError)")
    }
}
```

> If you have custom models or when Model have required fields which you have decided not to include in the response, you can create a `Codable` that conforms to the structure of the response data that you expect. From the previous example, the `Codable` can look like this

Reviewer: ```suggestion
If you have custom models or your Model has required fields that you have decided not to include in the response, you can create a `Codable` that conforms to the structure of the response data that you expect. From the previous example, the `Codable` would look like this
```
